src/EmailAccountRecovery.sol

```diff
@@ -0,0 +1,217 @@
+    uint8 public constant EMAIL_ACCOUNT_RECOVERY_VERSION_ID = 2;
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                    ERRORS                                  */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    error InvalidGuardianImplementation();
+    error InvalidProxyDeployment();
+    error GuardianNotDeployed();
+    error ProofVerificationFailed(string);
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                    Functions                               */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @notice Returns if the account to be recovered has already activated the controller (this
+    /// contract).
+    /// @dev This function is virtual and should be implemented by inheriting contracts.
+    /// @dev This function helps a relayer inactivate the guardians' data after the account
+    /// inactivates the controller (this contract).
+    /// @param recoveredAccount The address of the account to be recovered.
+    /// @return bool True if the account is already activated, false otherwise.
+    function isActivated(
+        address recoveredAccount
+    ) public view virtual returns (bool);
+
+    function acceptGuardian(
+        address guardian,
+        address recoveredAccount
+    ) internal virtual;
+
+    function processRecovery(
+        address guardian,
+        address recoveredAccount,
+        bytes32 recoveryDataHash
+    ) internal virtual;
+
+    /// @notice Completes the recovery process.
+    /// @dev This function must be implemented by inheriting contracts to finalize the recovery
+    /// process.
+    /// @param account The address of the account to be recovered.
+    /// @param completeCalldata The calldata for the recovery process.
+    function completeRecovery(
+        address account,
+        bytes memory completeCalldata
+    ) external virtual;
+
+    /// @notice Computes the address for guardian verifier contract using the CREATE2 opcode.
+    /// @dev This function utilizes the `Create2` library to compute the address. The computation
+    /// uses a provided account address to be recovered, account salt, and
+    /// the hash of the encoded ERC1967Proxy creation code concatenated with the encoded guardian verifier implementation
+    /// address and the initialization call data. This ensures that the computed address is
+    /// deterministic and unique per account salt.
+    /// @param guardianVerifierImplementation The address of the guardian verifier implementation.
+    /// @param recoveredAccount The address of the account to be recovered.
+    /// @param accountSalt A bytes32 salt value used to ensure the uniqueness of the deployed proxy address.
+    /// @param verifierInitData The initialization data for the guardian verifier.
+    /// @return address The computed address.
+    function computeGuardianVerifierAddress(
```

Comment:
Maybe we want to support non-upgradeable guardian verifiers, in which case this logic would change. Hmm let’s leave it for now, but will think more on this

Reply:
Leaning towards non-upgradeable guardians

Reply:
we were having similar conversations when designing the `EmailSigner`, and came to the conclusion we would remove upgradeability by default and only use it where it is needed (email signer would not be upgradeable, but the underlying verifier implementation is used would be)

For this PR, if you have time, would simplify and switch to non-upgradeable guardians

Reply:
https://github.com/zkemail/email-tx-builder/blob/main/packages/contracts/src/EmailSigner.sol

Reply:
I understand, but I would still want to use proxy-clone method for guardian verifiers, so that the deployment cost is minimal.

I can sure remove the upgradeability part here. Let me know if clones work according to you - https://docs.openzeppelin.com/contracts/4.x/api/proxy#Clones

Reply:
Yeah that's fair, Clones work

Reply:
I checked this further post our conversation on the call, the current implementation with ERC1967 is in fact not upgradeable by default, it's just a proxy, it become upgradeable if you were to use `ERC1967Upgrade` or the implementation contract ( that being guardian verifier) follow an upgradeable pattern. So it's safe to leave it as such